mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
+    if (type.isInteger(1)) {
+      rewriter.replaceOpWithNewOp<EmitCOp>(op, type, adaptor.getLhs(),
+                                           adaptor.getRhs());
+      return success();
+    }
+
+    Type arithmeticType = adaptIntegralTypeSignedness(type, true);
+
+    Value lhs = adaptValueType(adaptor.getLhs(), rewriter, arithmeticType);
+    Value rhs = adaptValueType(adaptor.getRhs(), rewriter, arithmeticType);
+
+    Value arithmeticResult = rewriter.template create<EmitCOp>(
+        op.getLoc(), arithmeticType, lhs, rhs);
+
+    Value result = adaptValueType(arithmeticResult, rewriter, type);
+
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+};
+
+template <typename ArithOp, typename EmitCOp, bool isUnsignedOp>
+class ShiftOpConversion : public OpConversionPattern<ArithOp> {
+public:
+  using OpConversionPattern<ArithOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArithOp op, typename ArithOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type type = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType, emitc::SignedSizeTType, emitc::SizeTType>(
+            type)) {
+      return rewriter.notifyMatchFailure(
+          op, "expected integer or size_t/ssize_t type");
+    }
+
+    if (type.isInteger(1)) {
+      return rewriter.notifyMatchFailure(op, "i1 type is not implemented");
+    }
+
+    Type arithmeticType = adaptIntegralTypeSignedness(type, isUnsignedOp);
+
+    Value lhs = adaptValueType(adaptor.getLhs(), rewriter, arithmeticType);
+    Value rhs = adaptValueType(adaptor.getRhs(), rewriter, arithmeticType);
+
+    // Add a runtime check for overflow
+    // This is an abuse of the size_t type since we're potentially using values
+    // below -1.
+    Value width;
+    Type sizeTType = (isa<IntegerType>(type)) ? arithmeticType
+                     : (isUnsignedOp)
+                         ? (Type)(emitc::SizeTType::get(op.getContext()))
+                         : (Type)(emitc::SignedSizeTType::get(op.getContext()));
+    if (isa<emitc::SignedSizeTType, emitc::SizeTType>(type)) {
+      Value eight = rewriter.create<emitc::ConstantOp>(
+          op.getLoc(), sizeTType, rewriter.getIndexAttr(8));
+      emitc::CallOpaqueOp sizeOfCall = rewriter.create<emitc::CallOpaqueOp>(
+          op.getLoc(), sizeTType, "sizeof", SmallVector<Value, 1>({eight}));
+      width = rewriter.create<emitc::MulOp>(op.getLoc(), sizeTType, eight,
+                                            sizeOfCall.getResult(0));
+    } else {
+      width = rewriter.create<emitc::ConstantOp>(
+          op.getLoc(), sizeTType,
+          rewriter.getIntegerAttr(sizeTType, type.getIntOrFloatBitWidth()));
+    }
+
+    Value oobTest;
+    Value excessCheck = rewriter.create<emitc::CmpOp>(
+        op.getLoc(), rewriter.getI1Type(), emitc::CmpPredicate::lt, rhs, width);
+    oobTest = excessCheck;
+    if (!isUnsignedOp) {
```

Comment:
The second operand is always interpreted as unsigned, even for https://mlir.llvm.org/docs/Dialects/ArithOps/#arithshrsi-arithshrsiop: `The first operand is interpreted as signed, and the second operand is interpreter as unsigned.`.